mono/utils/mono-threads-windows.c

```diff
@@ -18,6 +18,157 @@
+{
+	THREADS_INTERRUPT_DEBUG ("%06d - interrupt_apc () called", GetCurrentThreadId ());
+}
+
+void
+mono_win32_interrupt_wait (PVOID thread_info, HANDLE native_thread_handle, DWORD tid)
+{
+	request_interrupt (thread_info, native_thread_handle, WIN32_APC_INFO_PENDING_INTERRUPT_SLOT, interrupt_apc, tid);
+}
+
+static void CALLBACK
+abort_apc (ULONG_PTR param)
+{
+	THREADS_INTERRUPT_DEBUG ("%06d - abort_apc () called", GetCurrentThreadId ());
+
+	MonoThreadInfo *info = mono_thread_info_current_unchecked ();
+	if (info) {
+		// Check if pending interrupt is still relevant and current thread has not left alertable wait region.
+		// NOTE, can only be reset by current thread, currently running this APC.
+		gint32 win32_apc_info = mono_atomic_load_i32 (&info->win32_apc_info);
+		if (win32_apc_info & WIN32_APC_INFO_PENDING_ABORT_SLOT) {
+			// Check if current thread registered an IO handle when entering alertable wait (blocking IO call).
+			// No need for CAS on win32_apc_info_io_handle since its only loaded/stored by current thread
+			// currently running APC.
+			HANDLE io_handle = (HANDLE)info->win32_apc_info_io_handle;
+			if (io_handle != INVALID_HANDLE_VALUE) {
+				// In order to break IO waits, cancel all outstanding IO requests.
+				// Start to cancel IO requests for the registered IO handle issued by current thread.
+				// NOTE, this is NOT a blocking call.
+				CancelIo (io_handle);
+			}
+		}
+	}
+}
+
+// Attempt to cancel sync blocking IO on abort syscall requests.
+// NOTE, the effect of the canceled IO operation is unknown so the caller need
+// to close used resources (file, socket) to get back to a known state. The need
+// to abort blocking IO calls is normally part of doing a thread abort, then the
+// thread is going away meaning that no more IO calls will be issued against the
+// same resource that was part of the cancelation. Current implementation of
+// .NET Framework and .NET Core currently don't support the ability to abort a thread
+// blocked on sync IO calls, see https://github.com/dotnet/corefx/issues/5749.
+// Since there is no solution covering all scenarios aborting blocking syscall this
+// will be on best effort and there might still be a slight risk that the blocking call
+// won't abort (depending on overlapped IO support for current file, socket).
+static void
+suspend_abort_syscall (PVOID thread_info, HANDLE native_thread_handle, DWORD tid)
+{
+	request_interrupt (thread_info, native_thread_handle, WIN32_APC_INFO_PENDING_ABORT_SLOT, abort_apc, tid);
+
+#if G_HAVE_API_SUPPORT(HAVE_CLASSIC_WINAPI_SUPPORT)
+	// In case thread is blocked on sync IO preventing it from running above queued APC, cancel
+	// all outputstanding sync IO for target thread. If its not blocked on a sync IO request, below
+	// call will just fail and nothing will be canceled. If thread is waiting on overlapped IO,
+	// the queued APC will take care of cancel specific outstanding IO requests.
+	CancelSynchronousIo (native_thread_handle);
```

Comment:
I am seeing a problem where `suspend_abort_syscall` is called from `mono_threads_suspend_begin_async_suspend`. However, the target thread was already suspended in `mono_threads_suspend_begin_async_suspend`. This seems to cause the call to `CancelSynchronousIo` to block/hang forever. It seems like we want to make the call to `CancelSynchronousIo ` either before suspending the thread or after resuming it, similar to mechanism in `mono_thread_info_install_interrupt` which calls after resuming.